### PR TITLE
rewrite struct constructor section

### PR DESCRIPTION
**Pull request by WalterBright**
It was woefully incomplete. Copied a bunch from the recent class construction spec.

### CODE REVIEW

File: spec/struct.dd [hunk 1]
```diff
@@ -559,6 +609,124 @@ $(H2 $(LNAME2 disable_default_construction, Disabling Default Struct Constructio
         $(BEST_PRACTICE Disabling default construction is useful when the default value,
         such as `null`, is not acceptable.)
 
+
+$(H2 $(LNAME2 UnionConstructor, Union Constructors))
+
+        $(P Unions are constructed in the same way as structs.)
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply spec/struct.dd]: Try it and see :-) Redefining what it should do is a topic for another PR.

File: spec/struct.dd [hunk 2]
```diff
@@ -559,6 +609,124 @@ $(H2 $(LNAME2 disable_default_construction, Disabling Default Struct Constructio
...
+            void $(CODE_HIGHLIGHT opAssign)(A rhs) {}
+        }
+        struct S
+        {
+            A val;
+            this(int i)
+            {
+                val = A(i);  // val is initialized to the value of A(i)
+                val = A(2);  // rewritten to val.opAssign(A(2))
+            }
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply spec/struct.dd]: Specifying it that way would be redundant. All fields are statically initialized before the constructor gets control, so any value the field has is therefore readable and legal.

**Comment by WalterBright** [comment]
@andralex all comments addressed.

File: spec/struct.dd (hunk 1 above)
[the remark this answers is not in the document]
Comment (andralex) [reply spec/struct.dd]: Oi.

File: spec/struct.dd [hunk 3]
```diff
@@ -372,108 +378,122 @@ $(H2 $(LEGACY_LNAME2 Struct-Constructor, struct-constructor, Struct Constructors
...
+        }
+        ---
+        )
+        )
 
-$(P Also, member assignments inside loops are never considered constructors,
-even if it can be determined statically that the loop executes at most once.
-Example:)
+        $(LI It is illegal to refer to $(D this) implicitly or explicitly
+        prior to making a delegate constructor call.)
```
Comment (RazvanN7) [inline spec/struct.dd]: This needs an example. I'm not sure what you are referring to.

File: spec/struct.dd [hunk 4]
```diff
@@ -372,108 +378,122 @@ $(H2 $(LEGACY_LNAME2 Struct-Constructor, struct-constructor, Struct Constructors
...
-{
-    A a;
-    this(int x)
-    {
-        foreach (i; 0 .. x ? 0 : 1) a = i; // error, cannot assign
-    }
-}
-------
+        $(LI Once the delegate constructor returns, all fields are considered
+        constructed.)
```
Comment (RazvanN7) [inline spec/struct.dd]: I would point out the fact that the compiler does not enforce that the delegate constructor *actually* initializes the field.

File: spec/struct.dd [hunk 5]
```diff
@@ -493,24 +513,70 @@ constructor will have their `.init` values upon destruction.)
...
+            // Based on the definition, this creates a mutable object. But the
+            // created object cannot contain any mutable global data.
+            // Therefore the created object is unique.
+
+            this(int[] arr) immutable pure;
+            // Based on the definition, this creates an immutable object. But
+            // the argument int[] never appears in the created object so it
+            // isn't implicitly convertible to immutable. Also, it cannot store
+            // any immutable global data.
+            // Therefore the created object is unique.
```
Comment (RazvanN7) [inline spec/struct.dd]: I think this should be reconsidered : https://issues.dlang.org/show_bug.cgi?id=10368 . Consider : 

```d
immutable int i;

struct C
{
    int* p;

    this(int) immutable pure
    { p = &i; }

}
void main()
{
    ++*C(9).p; // changes `i`
}
```
One cannot reliably construct a mutable object with an `immutable pure` constructor.

File: spec/struct.dd [hunk 6]
```diff
@@ -559,6 +609,124 @@ $(H2 $(LNAME2 disable_default_construction, Disabling Default Struct Constructio
         $(BEST_PRACTICE Disabling default construction is useful when the default value,
         such as `null`, is not acceptable.)
 
+
+$(H2 $(LNAME2 UnionConstructor, Union Constructors))
+
+        $(P Unions are constructed in the same way as structs.)
+
+
+$(H2 $(LNAME2 field-init, Field initialization inside constructor))
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply spec/struct.dd]: I suggest that the delegate constructor paragraph from above (lines 401-488) be added at this section as a `Restrictions` section.

File: spec/struct.dd [hunk 7]
```diff
@@ -555,10 +622,161 @@ $(H2 $(LNAME2 disable_default_construction, Disabling Default Struct Constructio
...
+
+        $(P In a constructor body, if a delegate constructor is called,
+        all field assignments are considered assignments.
+        Otherwise, the first instance of field assignment is
+        its initialization, and assignments of the form `field = expression`
+        is treated as equivalent to `typeof(field)(expression)`.
```
Comment (RazvanN7) [inline spec/struct.dd]: "and assignments of the form `field = expression` **are** treated..."

File: spec/struct.dd [hunk 8]
```diff
@@ -555,10 +622,161 @@ $(H2 $(LNAME2 disable_default_construction, Disabling Default Struct Constructio
...
+        The values of fields may be read before initialization or construction
+        with a delegate constructor.
```
Comment (RazvanN7) [inline spec/struct.dd]: Except if the field is an aggregate declaration with default construction disabled. In that case, the compiler will (or should) issue an error.

File: spec/struct.dd [hunk 9]
```diff
@@ -555,10 +622,161 @@ $(H2 $(LNAME2 disable_default_construction, Disabling Default Struct Constructio
...
+                    default: break;
+                }
+                num = j;        // Error: `case` and `default` are also labels
+            }
+        }
+        ---
+        )
+
+        $(P If a field's type has disabled default construction, then it must be initialized
+        in the constructor.)
```
Comment (RazvanN7) [inline spec/struct.dd]: Please mention that using the field before initialization results in an error (or if it doesn't then it is a bug)

Comment (RazvanN7) [reply spec/struct.dd]: After this is merged : https://github.com/dlang/dmd/pull/8201 , the code in the original bug report provides a nice example which, in my opinion, should be present here.

File: spec/struct.dd (hunk 2 above)
[the remark this answers is not in the document]
Comment (RazvanN7) [reply spec/struct.dd]: Not if `A` has disabled default construction

File: spec/struct.dd (hunk 4 above)
Comment (RazvanN7) [reply spec/struct.dd]: What I mean is that this is valid code:
```d
struct A
{   
    int a, b, c;
    this(int a)
    {
        this.a = a;
    }
 
    this(int a, int b)
    {
        this(a);
    }
}
```

File: spec/struct.dd (hunk 5 above)
Comment (WalterBright) [reply spec/struct.dd]: It's a great point, but I am trying to document what is, not what should be. This should be addressed when dealing with 10368.

File: spec/struct.dd (hunk 6 above)
[the remark this answers is not in the document]
Comment (WalterBright) [reply spec/struct.dd]: Copy/pasting 87 lines of spec is not the best way to do things. Things should only be specified once.

File: spec/struct.dd (hunk 7 above)
Comment (WalterBright) [reply spec/struct.dd]: ok

File: spec/struct.dd (hunk 8 above)
Comment (WalterBright) [reply spec/struct.dd]: I'm not sure that change is right. In any case, again, here I am documenting what is, not what it should be.

File: spec/struct.dd (hunk 9 above)
Comment (WalterBright) [reply spec/struct.dd]: Again, I am documenting what is, not what should be. Just like we should not mix refactoring code with redesigning it.

**Pull request merged**